management/src/main/java/io/micronaut/management/endpoint/env/EnvironmentFilterSpecification.java

```diff
@@ -0,0 +1,117 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.management.endpoint.env;
+
+import io.micronaut.core.annotation.Nullable;
+
+import javax.validation.constraints.NotNull;
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class is passed to an instance of {@link EnvironmentEndpointFilter} (if one is defined) each time the {@link EnvironmentEndpoint}
+ * is invoked.
+ *
+ * @author Tim Yates
+ * @since 3.3.0
+ */
+public final class EnvironmentFilterSpecification {
+
+    private static final List<Pattern> LEGACY_MASK_PATTERNS = Stream.of("password", "credential", "certificate", "key", "secret", "token")
+            .map(s -> Pattern.compile(".*" + s + ".*", Pattern.CASE_INSENSITIVE))
+            .collect(Collectors.toList());
+
+    @Nullable
+    private final Principal principal;
+    private boolean allMasked;
+    private final List<Pattern> maskedPatterns = new ArrayList<>();
+
+    EnvironmentFilterSpecification(@Nullable Principal principal) {
+        this.principal = principal;
+        this.allMasked = true;
+    }
+
+    /**
+     * @return The current {@link Principal} that is making the request (if any)
+     */
+    public @Nullable Principal getPrincipal() {
+        return principal;
+    }
+
+    /**
+     * Turn on global masking. Items can be unmasked by adding their {@link Pattern} to {@link EnvironmentFilterSpecification#maskPatterns(Pattern...)}.
+     *
+     * @return itself for chaining calls.
+     */
+    public @NotNull EnvironmentFilterSpecification maskAll() {
+        allMasked = true;
+        return this;
+    }
+
+    /**
+     * Turn off global masking. Items can be masked by adding their {@link Pattern} to {@link EnvironmentFilterSpecification#maskPatterns(Pattern...)} (Pattern...)}.
+     *
+     * @return itself for chaining calls.
+     */
+    public @NotNull EnvironmentFilterSpecification maskNone() {
+        allMasked = false;
+        return this;
+    }
+
+    /**
+     * Adds patterns to the list of known mask patterns.  If the {@link EnvironmentFilterSpecification#maskAll()} flag is set
+     * then this list will be used for exceptions (values in clear-text).  If the {@link EnvironmentFilterSpecification#maskNone()}
+     * flag is set then this list will be those values that are masked.
+     *
+     * @param propertySourceNameMask one or more Patterns to match against property names for masking.
+     * @return itself for chaining calls.
+     */
+    public @NotNull EnvironmentFilterSpecification maskPatterns(Pattern... propertySourceNameMask) {
+        maskedPatterns.addAll(Arrays.asList(propertySourceNameMask));
```

Comment:
This would have no effect without `allMasked = false;` so that should probably be set here

Reply:
If `allMasked = true`, then this sets the patters that are allowed to be in the clear

https://github.com/micronaut-projects/micronaut-core/pull/6711/files#diff-4e7e5d0b9696543f54847a8d042130fc33e81143e23d6c414504d106ec0fa4c7R15

`maskPatterns` is probably a rubbish name...  But `toggleMask` was wrong too, as it's not a toggle, it's additive

Reply:
I think `maskPatterns` should define the patterns that are masked, thus only making it useful if not all keys are masked

Reply:
With a separate list for unmasked patterns, that only come into play if allMasked is true?

Or we could name it "excludedPatterns" which are excluded from whichever mode you're in... And change the method to exclude() ?

Reply:
That sounds OK to me. Need to support static keys as well

Reply:
Also the ability to supply predicates. Everything could be converted to predicates

Reply:
Switch to using `Predicate<String>` for all `exclusions`

To avoid `unchecked generic array creation for varargs parameter` warnings in user code (due to the `Predicate<String>` varargs), introduced `EnvironmentFilterNamePredicate` as a typealias

Also added a factory method to generate a predicate from a pattern and optional flags